### DISCUSSION TITLE
classes/class_@globalscope

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#81] [User-contributed notes]
# classes/class_@globalscope

Global scope constants and functions. Description: A list of global scope enumerated constants and built-in functions. This is all that resides in the globals, constants regarding error codes, keyc...

https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html

**Comment by JMVRy** [comment]
Note to anyone who does not know the C# equivalent to the GDScript `linear_to_db` or `db_to_linear` functions:

it's `Mathf.LinearToDb` or `Mathf.DbToLinear`. For some reason, the docs mention the GDScript version of the function here, but you'd have to go to [a different page](https://docs.godotengine.org/en/latest/tutorials/scripting/c_sharp/c_sharp_differences.html#full-list-of-equivalences) to find out. Weird.

**Comment by AM-Mikey** [comment]
The Property Usage Flags for Exported variables is PROPERTY_USAGE_STORAGE + PROPERTY_USAGE_EDITOR + PROPERTY_USAGE_SCRIPT_VARIABLE. I feel like this should be written down in the documentation somewhere.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/98478

**Comment by MaksymPy** [comment]
I don't know if this is correct, but the numbers fit.
<code>
        "name": "enum_class_property_name",
		"class_name": "&\"ClassName.EnumClassName\"",
		"type": TYPE_INT,
		"hint": PROPERTY_HINT_ENUM,
		"hint_string": "String1:0,String2:1,String3:2",
		"usage": PROPERTY_USAGE_CLASS_IS_ENUM + PROPERTY_USAGE_DEFAULT + PROPERTY_USAGE_SCRIPT_VARIABLE  #69638
</code>

**Comment by brantagames** [comment]
Here is a Desmos graph of the `ease` function I made: https://www.desmos.com/calculator/fyp62dbiw6.

**Comment by Stan-Stani** [comment]
`str_to_var()` needs better documentation on usage. 

For example, if you want to convert a vector string produced by `JSON.stringify()` to a vector 3,
`str_to_var("Vector3(3,3,3)")`.

**Comment by BitsAtPlayDev** [comment]
In the GDScript example code for `instance_from_id()`:
```gdscript
var drink = "water"

func _ready():
	var id = get_instance_id()
	var instance = instance_from_id(id)
	print(instance.foo) # Prints "water"
```
in the last line should be `instance.drink` not `instance.foo`

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/112486

**Comment by BitsAtPlayDev** [comment]
For the `wrap()`, `wrapf()` & `wrapi()` methods, it might be helpful to mention that `min` is _inclusive_, but `max` is _exclusive_. 

`max` being exclusive is hinted at in the `wrapi` code example, but I think stating it explicitly could prevent any potential confusion.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/107166

**Comment by PangolinMontanari** [comment]
The documentation for `PROPERTY_HINT_TYPE_STRING` is misleading, since it makes heavy use of the `%` operator; this is invalid since hint strings need to be constant, which is not possible when using the `%` operator.

**Reply by Selkcip** [reply]
It's not clear how the hintString variable is intended to be used for GD or C#. At least for C# I was able to hard code the enum values. For an array of Strings (Variant.Type.String = 4) as FilePaths (PropertyHint.FilePath = 44):
`[Export(PropertyHint.TypeString, "4/44:*.tscn")]
private string[] _testPaths = [];`
The enums values are explicitly defined so hopefully they're consistent between releases.

**Comment by Varonex** [comment]
I wonder why `weakref(obj: Variant)` is marked as returning a `Variant`, whereas (to my understanding), it returns a `Weakref` object containing the instance, like `Weakref<T> weakref<T>(Obj: T)` and not `T weakref<T>(Obj: T)`. Of course this cannot be described in GDScript yet.

**Comment by Johnsn0w** [comment]
I'm assuming the Properties table at the top of the page is meant to be two columns of "Property Name" and "Description", as opposed to two duplicate columns of "Property Name". 
If it's a mistake then this is a bug report, if it's intentional then this is a suggestion.

**Reply by JMVRy** [reply]
@Varonex Github will give you a notification if anyone replies to the discussion. I've been getting notifications for every single comment here. If you want to stop receiving notifications for it, there's a button on the right to unsubscribe from this discussion thread.

**Comment by Sithoid** [comment]
Another note on linear_to_db: it doesn't validate the input range. This is actually useful because it lets you create a volume slider that goes up to 150% or 200% (input values 1.5 and 2.0, respectively), which can be handy for voice chats.

**Reply by Calinou** [reply]
Values above 1.0 are technically legal, but they can introduce distortion (clipping) if the source signal is loud enough.

**Comment by pumguy** [comment]
"Note that making a full-fledged property array requires boilerplate code involving Object._get_property_list()."
Please provide corresponding "boilerplate code" either here or below Object._get_property_list()

**Comment by admin** [comment]
*Originally posted by [**thiscris**](https://github.com/thiscris) on 2026-05-30 08:22:55 UTC*

> print_rich() supports more BBcode tags than the ones listed. [table] and [img] at least. Haven't done a complete test.

**Reply by admin** [reply]
*Originally posted by [**Calinou**](https://github.com/Calinou) on 2026-06-09 15:22:52 UTC*

> All BBCode tags technically work in the Output panel, but not all of them can be converted to ANSI escape codes for terminal display. Neither `[table]` nor `[img]` are supported for that purpose.
>
> It's technically possible to add ANSI escape code conversion for these tags (using [sixel](https://en.wikipedia.org/wiki/Sixel) or [kitty image protocol](https://sw.kovidgoyal.net/kitty/graphics-protocol/) for images), but this should be discussed in a [proposal](https://github.com/godotengine/godot-proposals).
>
> I've opened a proposal for named `[url]` support:
>
> - https://github.com/godotengine/godot-proposals/issues/14992

**Comment by greenfox1505** [comment]
The example `lerp_angle(` is extremely unhelpful. The result is indistinguishable from `lerp(`, which makes it seem pointless.

The example can be fixed by choosing two angles whose shortest path is through 0deg.

```gdscript
extends Sprite
var elapsed = 0.0
func _process(delta):
	var min_angle = deg_to_rad(45.0)
	var max_angle = deg_to_rad(315.0)
	rotation = lerp_angle(min_angle, max_angle, elapsed)
	elapsed += delta
```

A more complex example to more explicitly show the difference interactively.

```gdscript
extends Sprite
@export use_angle:=true
var elapsed = 0.0
func _process(delta):
	var min_angle = deg_to_rad(45.0)
	var max_angle = deg_to_rad(315.0)
	if use_angle:
		rotation = lerp_angle(min_angle, max_angle, elapsed)
	else:
		rotation = lerp(min_angle, max_angle, elapsed)
	elapsed += delta
	elapsed = wrapf(elapsed,0.0,1.0)
```
